### DISCUSSION TITLE
Conditional Looping in Guidance Program

### DISCUSSION THREAD

**Discussion by rickbraddy-pharma** [guidance-ai/guidance#191] [Q&A] [answered]
I am implementing a ReAct style Guidance program that calls out to my Tools for search, math, etc.  I'm able to successfully call tools (functions) now during a chain-of-thought process in the program; however, since the program flows linearly to the end, and my prompts are created in the User prompt space and generation must take place in Assistant for GPT models, how can I put the program into a loop to call as many tool functions as needed, until it reaches its "Final Answer"?

**Comment by AbdullahWasTaken** [comment]
I think you can try using a geneach block with a stop parameter set to "Final Answer". Although you can try to handle control flow using guidance, I have found it best to write control flow in python and use partials if possible.

You can refer to this article written by QuangBK [A Simple Agent With Guidance and Local LLM
](https://betterprogramming.pub/a-simple-agent-with-guidance-and-local-llm-c0865c97eaa9) implementing ReAct agent using guidance, but uses text completion models (not chat models)

**Comment by chrisfentiman** [comment, marked as answer]
The above is a great example, but here is some code to save people time and effort searching for answers.

```
{{#geneach 'some-var' stop=False}}
... do something
{{#if (some conditional)}}{{break}}{{/if}}
{{/geneach}}
```

You can use a `geneach` to continuously loop until your condition is met. Once met, you can use the break provided by guidance to break from the loop. This will then go ahead and execute the rest of your prompt as expected.

**Reply by rickbraddy-pharma** [reply]
Extremely helpful
